src/lib.rs

```diff
@@ -723,8 +551,281 @@ pub unsafe trait Unaligned {
+}
+
+safety_comment! {
+    // `NonZeroXxx` is `AsBytes`, but not `FromBytes`.
+    //
+    /// SAFETY:
+    /// `NonZeroXxx` has the same layout as its associated primitive. Since it
+    /// is the same size, this guarantees it has no padding - integers have no
+    /// padding, and there's no room for padding if it can represent all of the
+    /// same values except 0.
+    ///
+    /// TODO(https://github.com/rust-lang/rust/pull/104082): Cite documentation
+    /// that layout is the same as primitive layout.
+    unsafe_impl!(NonZeroU8: AsBytes);
+    unsafe_impl!(NonZeroI8: AsBytes);
+    unsafe_impl!(NonZeroU16: AsBytes);
+    unsafe_impl!(NonZeroI16: AsBytes);
+    unsafe_impl!(NonZeroU32: AsBytes);
+    unsafe_impl!(NonZeroI32: AsBytes);
+    unsafe_impl!(NonZeroU64: AsBytes);
+    unsafe_impl!(NonZeroI64: AsBytes);
+    unsafe_impl!(NonZeroU128: AsBytes);
+    unsafe_impl!(NonZeroI128: AsBytes);
+    unsafe_impl!(NonZeroUsize: AsBytes);
+    unsafe_impl!(NonZeroIsize: AsBytes);
+}
+
+safety_comment! {
+    /// SAFETY:
+    /// The Rust compiler reuses `0` value to represent `None`, so
+    /// `size_of::<Option<NonZeroXxx>>() == size_of::<xxx>()`; see `NonZeroXxx`
+    /// documentation.
+    ///
+    /// TODO(https://github.com/rust-lang/rust/pull/104082): Cite documentation
+    /// for layout guarantees.
+    unsafe_impl!(Option<NonZeroU8>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroI8>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroU16>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroI16>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroU32>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroI32>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroU64>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroI64>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroU128>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroI128>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroUsize>: FromBytes, AsBytes);
+    unsafe_impl!(Option<NonZeroIsize>: FromBytes, AsBytes);
+}
+
+safety_comment! {
+    /// SAFETY:
+    /// TODO
+    ///
+    /// TODO(https://github.com/rust-lang/rust/pull/104081): Cite guaranteed
+    /// size and alignment.
+    unsafe_impl!(T: ?Sized => FromBytes for PhantomData<T>);
+    unsafe_impl!(T: ?Sized => AsBytes for PhantomData<T>);
+    unsafe_impl!(T: ?Sized => Unaligned for PhantomData<T>);
+}
+safety_comment! {
+    /// SAFETY:
+    /// `Wrapping<T>` is guaranteed to have the same layout as `T`.
+    unsafe_impl!(T: FromBytes => FromBytes for Wrapping<T>);
+    unsafe_impl!(T: AsBytes => AsBytes for Wrapping<T>);
+    unsafe_impl!(T: Unaligned => Unaligned for Wrapping<T>);
+}
+
+safety_comment! {
+    // `MaybeUninit<T>` is `FromBytes`, but never `AsBytes` since it may contain
+    // uninitialized bytes.
+    //
+    /// SAFETY:
+    /// `MaybeUninit<T>` has no restrictions on its contents.
+    unsafe_impl!(T => FromBytes for MaybeUninit<T>);
```

Comment:
This should also implement `Unaligned` for `MaybeUninit<T>` where `T: Unaligned`.

Reply:
Agreed, but I'm gonna follow up on that in a subsequent PR. For the time being, I want this change to just be a refactor and adding safety arguments, not actually adding new impls.